### ISSUE TITLE
Should not create sync-conflict file if an identical sync-conflict file already exists

### ISSUE THREAD

**Issue by canton7** [syncthing/syncthing#2619]
I've got a situation where I've got a file with two conflicts. Those two conflicted files have the same sha1sum, which does differ from the original file.

When creating a conflict file, do we ensure that that file doesn't already exist as a conflict file from an earlier point in time?

 I can see how this situation could come about: Device B connect to Device A, sees that its copy of the file differs from Device A's, and creates a conflict, letting Device A's file win. Device C (which has the same copy of the file as Device A) does exactly the same thing later on. Afterwards, Devices B and C synchronize, and share their two copies of the conflict. 

If Device C first looks to see whether a conflicted copy of that file already exists (it may have to check to see whether it exists but hasn't been synchronized yet?) then this wouldn't happen, I think?

**Comment by canton7** [comment]
There may even be an argument for saying that if more devices end up having the losing copy of the conflict (as indicated by many devices creating a conflict, then seeing that it exists already), that losing copy should become the winning copy, by majority vote?

**Comment by canton7** [comment]
Maybe related to #2584?

**Comment by calmh** [comment]
> When creating a conflict file, do we ensure that that file doesn't already exist as a conflict file from an earlier point in time?

The file name used contains the current date and time at the time of the conflict resolution, to one second precision. I don't think this can ever collide?

**Comment by canton7** [comment]
It's not collisions I'm thinking about: almost the opposite. If we want to create a conflict file, but someone else already created a conflict file with the same contents at an earlier point in time, there's no point in us re-creating it. 

In a couple of messy situations I've had 3 or 4 conflict files for a given file, all with exactly the same contents.

**Comment by calmh** [comment]
Ah, right. That would require some special magic for avoiding to create conflict files with the same contents as another conflict file for the same base file... It's rather niche?

**Comment by canton7** [comment]
Well as I tried to describe above, I think it's got a reasonable chance if happening when there's a conflict in a setup with three or more devices. I've seen it without about a third of all conflicts I've seen, so i'm reasonably sure it's not too niche. 

Devices A and B have version X of the file, and device C has version Y. If A and C sync and version Y "wins", we'll get a sync-conflict file with version X. If devices A and C sync, I think we'll get another sync-conflict file with version X?

If that doesn't seem like probably behaviour I'll do some more testing in the new year, see if I can reliably repro.

**Comment by calmh** [comment]
Yeah, no, I see it happening for sure, I mean the logic itself is a bit niche. Opening and checking all existing conflict files for matches can also be quite slow/intense (if they are large), and we could potentially be close to losing data if the already existing conflict file is the tenth oldest and we're set to allow a maximum of ten. Not doing the conflict copy then leaves the (recent-ish) data in risk of being destroyed soon when it shouldn't. But that's even more of a corner case...

**Comment by canton7** [comment]
I may be betraying my ignorance here, but since all conflict files are normal tracked files with block hashes and all the rest of it, wouldn't the check be cheap? We've already hashed the files. 

Agreed on the potentially losing data, but taking the approach of deleting the old conflict file rather than not creating a new one should fix that.

**Comment by canton7** [comment]
Actually I may not have thought that last paragraph through...

**Comment by calmh** [comment]
No you're probably right. You'd need to iterate over the files in the database that have the expected prefix (`$folderdir/$actualFilename.sync-conflict-`) and check their block hashes. If you find a match, you could simply rename the old conflict file to a newer name (and touch it), and that would be it.

**Comment by canton7** [comment]
I'll have to think through a scenario where different devices keep doing this for the same file when I'm a bit more awake...

**Comment by nrm21** [comment]
> Devices A and B have version X of the file, and device C has version Y. If A and C sync and version Y "wins", we'll get a sync-conflict file with version X. If devices A and **B** sync, I think we'll get another sync-conflict file with version X?

If version Y were to win the A <=> C Sync, then why would it not also win the A <=> B sync?  After all version X is identical on  A and B right?  That means the same timestamp as well so if Y wins on one it will win on both.

**Comment by canton7** [comment]
> If version Y were to win the A <=> C Sync, then why would it not also win the A <=> B sync?

It will, and that's exactly what I described. 

> If devices A and B sync, I think we'll get another sync-conflict file with version X

We get a sync-conflict file containing version X because version Y won.
